parachain/src/primitives.rs

```diff
@@ -0,0 +1,221 @@
+// #360
+struct TrailingZeroInput<'a>(&'a [u8]);
+impl<'a> codec::Input for TrailingZeroInput<'a> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		Ok(None)
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		let len = into.len().min(self.0.len());
+		into[..len].copy_from_slice(&self.0[..len]);
+		for i in &mut into[len..] {
+			*i = 0;
+		}
+		self.0 = &self.0[len..];
+		Ok(())
+	}
+}
+
+/// Format is b"para" ++ encode(parachain ID) ++ 00.... where 00... is indefinite trailing
+/// zeroes to fill AccountId.
+impl<T: Encode + Decode + Default> AccountIdConversion<T> for Id {
+	fn into_account(&self) -> T {
+		(b"para", self).using_encoded(|b|
+			T::decode(&mut TrailingZeroInput(b))
+		).unwrap_or_default()
+	}
+
+ 	fn try_from_account(x: &T) -> Option<Self> {
+		x.using_encoded(|d| {
+			if &d[0..4] != b"para" { return None }
+			let mut cursor = &d[4..];
+			let result = Decode::decode(&mut cursor).ok()?;
+			if cursor.iter().all(|x| *x == 0) {
+				Some(result)
+			} else {
+				None
+			}
+		})
+	}
+}
+
+/// Which origin a parachain's message to the relay chain should be dispatched from.
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+#[repr(u8)]
+pub enum ParachainDispatchOrigin {
+	/// As a simple `Origin::Signed`, using `ParaId::account_id` as its value. This is good when
+	/// interacting with standard modules such as `balances`.
+	Signed,
+	/// As the special `Origin::Parachain(ParaId)`. This is good when interacting with parachain-
+	/// aware modules which need to succinctly verify that the origin is a parachain.
+	Parachain,
+	/// As the simple, superuser `Origin::Root`. This can only be done on specially permissioned
+	/// parachains.
+	Root,
+}
+
+impl sp_std::convert::TryFrom<u8> for ParachainDispatchOrigin {
+	type Error = ();
+	fn try_from(x: u8) -> core::result::Result<ParachainDispatchOrigin, ()> {
+		const SIGNED: u8 = ParachainDispatchOrigin::Signed as u8;
+		const PARACHAIN: u8 = ParachainDispatchOrigin::Parachain as u8;
+		Ok(match x {
+			SIGNED => ParachainDispatchOrigin::Signed,
+			PARACHAIN => ParachainDispatchOrigin::Parachain,
+			_ => return Err(()),
+		})
+	}
+}
+
+/// A message from a parachain to its Relay Chain.
+#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(
+	any(feature = "std", feature = "wasm-api"),
+	derive(sp_runtime_interface::pass_by::PassByCodec,
+))]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct UpwardMessage {
+	/// The origin for the message to be sent from.
+	pub origin: ParachainDispatchOrigin,
+	/// The message data.
+	pub data: Vec<u8>,
+}
+
+/// Validation parameters for evaluating the parachain validity function.
+// TODO: balance downloads (https://github.com/paritytech/polkadot/issues/220)
+#[derive(PartialEq, Eq, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Encode))]
+pub struct ValidationParams {
```

Comment:
```suggestion
#[derive(PartialEq, Eq, Encode, Decode)]
#[cfg_attr(feature = "std", derive(Debug))]
pub struct ValidationParams {
```

It would be helpful if coding on this struct were always bidirectional: that would let me encode it into storage (under a constant key) so that pallets could publish things like the max code size and whether or not code upgrades are currently allowed.

Alternate suggestion, in case `BlockData` or `HeadData` are large: factor out something like

```rust
#[derive(PartialEq, Eq, Encode, Decode, Clone, Copy)]
#[cfg_attr(feature = "std", derive(Debug))]
pub struct ValidationFunctionParams {
pub max_code_size: u32,
	/// The maximum head-data size permitted, in bytes.
	pub max_head_data_size: u32,
	/// The current relay-chain block number.
	pub relay_chain_height: RelayChainBlockNumber,
	/// Whether a code upgrade is allowed or not, and at which height the upgrade
	/// would be applied after, if so. The parachain logic should apply any upgrade
	/// issued in this block after the first block
	/// with `relay_chain_height` at least this value, if `Some`. if `None`, issue
	/// no upgrade.
	pub code_upgrade_allowed: Option<NonZero<RelayChainBlockNumber>>,
}
```

That's all the data I'm really interested in forwarding for pallet use; this struct could then become a public member of `ValidationParams`.

Reply:
> It would be helpful if coding on this struct were always bidirectional: that would let me encode it into storage (under a constant key) so that pallets could publish things like the max code size and whether or not code upgrades are currently allowed.

Yes, but at the cost of larger runtime code size.  You can build it with the `std` feature to get `Encode` functionality.

Reply:
Would you be open to the alternate suggestion, breaking out `ValidationFunctionParams`? I'm already doing that in the Cumulus branch; it's not hard, but it would be easier to just copy out the struct directly. 

https://github.com/paritytech/cumulus/pull/70/files#diff-11f21b7f145f98241270cf825e0aaab5R95-R122

Reply:
I think this stuff should be kept locally to Cumulus.

Reply:
Can you elaborate on why a Cumulus chain would want to put this in storage? These parameters will change for the execution of every block and cannot be relied on beyond a single block's execution.

The only thing a Cumulus should need to store to process a pending upgrade is what the inner value of `code_upgrade_allowed: Some` was when the upgrade was initiated. Then the next time you get a `ValidationFunctionParams` with `relay_chain_height >= scheduled_at`, you apply the code change at the end of the block by writing to `:code` storage.

Reply:
> Can you elaborate on why a Cumulus chain would want to put this in storage?

The main goal is to propagate the values beyond Cumulus, so that a pallet can use them. It enables the pallet to publish methods like `can_set_code() -> bool` or `max_code_size() -> u32`, but more importantly, it lets the pallet generate nice errors like this:

```rust
pub fn set_code(origin, validation_function: ValidationFunction) {
    let vfp = Self::validation_function_params();
    ensure!(validation_function.len() <= vfp.max_code_size as usize, Error::<T>::TooBig);
    let apply_block = vfp.code_upgrade_allowed.ok_or(Error::<T>::ProhibitedByPolkadot)?;
    ...
}
```

To a user, that's a much nicer way to know that this is the wrong time or the wrong size for a validation function upgrade than panicing during the Cumulus `validate_block` function.

I spoke with @bkchr earlier this afternoon, and he explained why literally putting it into storage is a bad idea (it messes with the storage root). We came up with workaround which still ends up going through the storage interface, even though it doesn't actually modify the storage; for that reason, it would be still good to derive `Encode`, `Decode` on the struct. 

> Is it a question of putting it in the ephemeral storage... I think you only need to put code_upgrade_allowed in such ephemeral storage.

Essentially, yes, that's why we're using it. We do use all three fields, though:

- `max_code_size`: so that we can return an error instead of panicing if submitted code is too big
- `relay_chain_height`: so we know when to set `:code` given a previously scheduled validation function upgrade
- `code_upgrade_allowed`: so we know when to schedule a validation function upgrade

> I wrote this code with the understanding that Cumulus would have an inherent type with the `ValidationFunctionParams`

I'm also looking into how to implement that. However, even if the upgrade pallet gets the validation function params from an inherent instead of by directly querying a special storage key, it's not yet clear to me how the inherent would get access to them except via the magic storage key.

Reply:
Additionally: it would be helpful to add a field to `ValidationParams` / `ValidationFunctionParams`. It could be as simple as 

```rust
expect_upgrade_this_block: bool,
```

Rationale: we can't use `code_upgrade_allowed` to validate all inserts to `:code`. Obviously the upgrade pallet will be well-behaved in this regard, but not all users will discover and use that pallet, so we should guard against errors in DIY user code.

Scenario:

- Block 1000: `can_set_code == Some(1500)`. User sets some code, which is returned in the `ValidationResult`.
- Blocks 1001-1499: `can_set_code == None`
- Block 1500: `can_set_code == None || can_set_code == Some(2000)`. I don't know which condition actually applies, but neither value indicates that it is legal to set `:code` on this block, right now. 

Given `expect_upgrade_this_block`, we can inspect all writes to `:code` and ensure that they are only performed when legal. Without it, we have to either store the scheduled upgrade block within `validate_block` (which breaks the storage root), or skip the check entirely.

Reply:
Ok, I will try to address these points one-by-one.

> Essentially, yes, that's why we're using it. We do use all three fields, though:
> 
> * `max_code_size`: so that we can return an error instead of panicing if submitted code is too big
> * `relay_chain_height`: so we know when to set `:code` given a previously scheduled validation function upgrade
> * `code_upgrade_allowed`: so we know when to schedule a validation function upgrade

I would still recommend to create a separate type for these fields, as we will add more stuff to `ValidationFunctionParams` including incoming messages and other fields you would not want to necessarily include in the ephemeral storage.

> Additionally: it would be helpful to add a field to `ValidationParams` / `ValidationFunctionParams`
 
`expect_upgrade_this_block` is redundant which could be a cause of confusion or bugs later on. I would prefer to avoid it. Specifically, this statement above is not correct:

> Block 1500: `can_set_code == None || can_set_code == Some(2000)`. I don't know which condition actually applies, but neither value indicates that it is legal to set `:code` on this block, right now.

When `can_set_code` is `Some`, you can schedule an upgrade to be applied at/after that block. The implementation accounts for the case where an upgrade is applied and a new one is signaled at the same time here: https://github.com/paritytech/polkadot/pull/918/files#diff-e21f8096708b22fcdbd66ebc74f0821aR1038

> > I wrote this code with the understanding that Cumulus would have an inherent type with the `ValidationFunctionParams`
> 
> I'm also looking into how to implement that. However, even if the upgrade pallet gets the validation function params from an inherent instead of by directly querying a special storage key, it's not yet clear to me how the inherent would get access to them except via the magic storage key.

The inherent should provide the value for the ephemeral magic storage key, right? And then the inherent is constructed via a `ProvideInherent` implementation, whose `InherentData` contains a `ValidationFunctionParams` which is provided by the `polkadot-collator` pipeline to Cumulus' authorship process.

Reply:
> When `can_set_code` is `Some`, you can schedule an upgrade to be applied at/after that block. When an upgrade is scheduled, you should put it into code as soon as `relay_chain_height >= scheduled_at`, regardless of value of `can_set_code`.

Yes, I understand that; that's what the pallet does. However, there are three areas in the code in play:

- Polkadot
- cumulus runtime (`validate_block`)
- pallet-parachain-upgrade

When an upgrade is scheduled, both polkadot and the pallet store the current value of `can_set_code`, and apply the code change at the appropriate time. However, the cumulus runtime is stateless; it doesn't get to add anything persistent to storage. 

That's all fine, assuming that the only way that users ever attempt parachain upgrades is via our pallet. However, that's not an entirely reasonable assumption. Someone out there, eventually, is likely to want to do it themselves for some reason. 

The purpose of `expect_upgrade_this_block` is to enable `validate_block` to know when an upgrade is legal, given that it can't store its own data in storage, and shouldn't peek into the pallet's storage (because we can't assume that our pallet will be used). If it exists, then `validate_block` can fail when a user attempts to set `:code` at the wrong time. If it doesn't exist, then it cannot perform that validation. If that validation is already performed by polkadot, then we can skip this. However, if Polkadot is not performing that validation check, then it is possible for an end user to write a pallet which just updates `:code` whenever it feels like. 

> And then the inherent is constructed via a `ProvideInherent` implementation, whose `InherentData` contains a `ValidationFunctionParams` which is provided by the `polkadot-collator` pipeline to Cumulus' authorship process.

Thanks! This gives me a lot of entrypoints to research to figure out how to do this.

Reply:
There are a lot of ways someone can write an incorrect parachain or runtime. Even if we added this `expect_upgrade_this_block`, there is still a possibility that someone writes a pallet that doesn't care about `can_upgrade_at` and writes to `:code` anyway. Or writes the wrong code to `:code`. Or doesn't write anything to`:code` when it's supposed to. We can't address these issues, so are we supposed to throw away this PR?

As long as Cumulus can expose a reasonable entry-point for users to perform parachain code upgrades, then responsible parachain authors and auditors will (need to) ensure that it is used correctly.